### PR TITLE
Adding repair step to fix missing etags - fixes #12172

### LINKED ISSUE #12172: Deal with files without ETag in filecache
about ownCloud version 7.0.3, installed from openSUSE packages

Because of errors from the past, production ownCloud may contain entries with empty ETag in the table ocfilecache, like this:
```
MariaDB [oc45]> select * from oc_filecache where path like '%kraftkleber.pdf';  
+--------+---------+-----------------------+----------------------------------+--------+-----------------+----------+----------+--------+------------+-----------+------+------------------+---------------+-------------+
| fileid | storage | path                  | path_hash                        | parent | name            | mimetype | mimepart | size   | mtime      | encrypted | etag | unencrypted_size | storage_mtime | permissions |
+--------+---------+-----------------------+----------------------------------+--------+-----------------+----------+----------+--------+------------+-----------+------+------------------+---------------+-------------+
|   8156 |       9 | files/kraftkleber.pdf | 617d4c2eee8e7020dc8c0a98091de1e6 |   2185 | kraftkleber.pdf |       15 |       10 | 177714 | 1362504934 |         0 |      |                0 |             0 |          31 |
+--------+---------+-----------------------+----------------------------------+--------+-----------------+----------+----------+--------+------------+-----------+------+------------------+---------------+-------------+
```
In my particular case, this file is not directly my file but was shared back in the ownCloud 6 days. 

Unfortunately this file breaks syncing, it causes an error "No E-Tag received from server, check Proxy/Gateway", as reported in https://github.com/owncloud/mirall/issues/1291

In owncloud.log, I can find this entry, no idea, if related:
```
{"reqId":"545cd28b4e979","app":"OC\\Files\\Cache\\Scanner","message":"!!! No reuse of etag for 'files' !!! \ncache: Array\n(\n    [fileid] => 12\n    [storage] => local::\/var\/owncloud\/data\/kf\/\n    [path] => files\n    [parent] => -1\n    [name] => files\n    [mimetype] => httpd\/unix-directory\n    [mimepart] => httpd\n    [size] => 1932889959\n    [mtime] => 1415369317\n    [storage_mtime] => 1415369317\n    [encrypted] => \n    [unencrypted_size] => 0\n    [etag] => 545cd26548e77\n    [permissions] => 31\n)\n \ndata: Array\n(\n    [mimetype] => httpd\/unix-directory\n    [mtime] => 1415369355\n    [size] => -1\n    [etag] => 545cd28b4e781\n    [storage_mtime] => 1415369355\n    [permissions] => 31\n)\n","level":0,"time":"2014-11-07T14:09:15+00:00","method":"GET","url":"\/owncloud\/remote.php\/webdav\/Shared\/kraftkleber.pdf"}
```

The server should be able to handle this case and somehow fix the empty etag, even if it means that the file is redownloaded.

### PR DESCRIPTION
**Pull request by DeepDiver1975**
fixes #12172

@dragotin @PVince81 @icewind1991 

From the pure sync algorithm pov it is sufficient to fill the etag with any random string which is not empty.

### CODE REVIEW

File: lib/private/repair.php [hunk 1]
```diff
@@ -79,7 +80,8 @@ public static function getRepairSteps() {
 			new RepairMimeTypes(),
 			new RepairLegacyStorages(\OC::$server->getConfig(), \OC_DB::getConnection()),
 			new RepairConfig(),
-			new AssetCache()
+			new AssetCache(),
+			new FillETags(\OC_DB::getConnection())
```
Comment (nickvergessen) [inline lib/private/repair.php]: add trailing `,` so the next repair step does not requrie changing two lines again?

**Comment by icewind1991** [comment]
Looks good

File: lib/repair/filletags.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace OC\Repair;
+
+use Doctrine\DBAL\Query\QueryBuilder;
+use OC\Hooks\BasicEmitter;
+
+class FillETags extends BasicEmitter implements \OC\RepairStep {
+
+	/** @var \OCP\IConfig */
+	protected $config;
```
Comment (LukasReschke) [inline lib/repair/filletags.php]: Uneeded?

File: lib/repair/filletags.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		$this->connection = $connection;
+	}
+
+	public function getName() {
+		return 'Generate ETags for file where no ETag is present.';
+	}
+
+	public function run() {
+		$qb = $this->connection->createQueryBuilder();
+		$qb->update('`*PREFIX*filecache`')
```
Comment (MorrisJobke) [inline lib/repair/filletags.php]: Interesting. Is this new?

Comment (DeepDiver1975) [inline lib/repair/filletags.php]: querybuilder? no - part of dbal for a long time afaik

**Comment by DeepDiver1975** [comment]
refs https://github.com/owncloud/core/issues/12502

File: lib/repair/filletags.php (hunk 3 above)
Comment (MorrisJobke) [inline lib/repair/filletags.php]: I never have seen something like this in our codebase before xD

Comment (LukasReschke) [inline lib/repair/filletags.php]: Because we're fans of string concatenation.

**Comment by MorrisJobke** [comment]
Works as intended. I deleted some etags and run the repair step :+1:

**Pull request merged**